apps/openmw/mwworld/containerstore.hpp

```diff
@@ -187,96 +200,143 @@ namespace MWWorld
+        mutable PtrType mPtr;
+
+        typename Iterator<ESM::Potion>::type mPotion;
+        typename Iterator<ESM::Apparatus>::type mApparatus;
+        typename Iterator<ESM::Armor>::type mArmor;
+        typename Iterator<ESM::Book>::type mBook;
+        typename Iterator<ESM::Clothing>::type mClothing;
+        typename Iterator<ESM::Ingredient>::type mIngredient;
+        typename Iterator<ESM::Light>::type mLight;
+        typename Iterator<ESM::Lockpick>::type mLockpick;
+        typename Iterator<ESM::Miscellaneous>::type mMiscellaneous;
+        typename Iterator<ESM::Probe>::type mProbe;
+        typename Iterator<ESM::Repair>::type mRepair;
+        typename Iterator<ESM::Weapon>::type mWeapon;
+
+        ContainerStoreIteratorBase (ContainerStoreType container);
+        ///< End-iterator
+
+        ContainerStoreIteratorBase (int mask, ContainerStoreType container);
+        ///< Begin-iterator
+
+        // construct iterator using a CellRefList iterator
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Potion>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Apparatus>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Armor>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Book>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Clothing>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Ingredient>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Light>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Lockpick>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Miscellaneous>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Probe>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Repair>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Weapon>::type);
+
+        template<class T>
+        void copy (const ContainerStoreIteratorBase<T>& src);
+        
+        void incType ();
+        
+        void nextType ();
+
+        bool resetIterator ();
+        ///< Reset iterator for selected type.
+        ///
+        /// \return Type not empty?
+
+        bool incIterator ();
+        ///< Increment iterator for selected type.
+        ///
+        /// \return reached the end?
 
         public:
+            template<class T>
+            ContainerStoreIteratorBase (const ContainerStoreIteratorBase<T>& other)
+            {
+                char CANNOT_CONVERT_CONST_ITERATOR_TO_ITERATOR[IsConvertible<T, PtrType, void>::value ? 1 : -1];
```

Comment:
Yeah, done. Ported simple loops that required very small change to code with no conversions (if you consider RVO / NRVO).

For example:
```c++
void Actor::block(const MWWorld::Ptr &ptr) const
{
    MWWorld::InventoryStore& inv = getInventoryStore(ptr);
    MWWorld::ContainerStoreIterator shield = inv.getSlot(MWWorld::InventoryStore::Slot_CarriedLeft);
    /* ... */
}
```
Changing the iterator to const version would trigger a conversion that can't be optimized by the compiler with RVO/NRVO. The right thing seems to create a const version for the method getSlot, but the internal representation is of course non const, then the conversion will be triggered anyway.